keps/sig-api-machinery/0030-storage-migration.md

```diff
@@ -256,28 +249,28 @@ Before upgrading or downgrading the cluster, the cluster administrator must run
 migrations have completed. Otherwise the apiserver can crash, because it cannot
 interpret the serialized data in etcd. To mitigate, the cluster administrator
 can rollback the apiserver to the old version, and wait for the migration to
-complete. Even if the apiserver does not crash after upgrading or downgrading,
-the `migration` objects are not accurate anymore, because the default storage
-versions might have changed after upgrading or downgrading, but no one
-increments the `migration.spec.generation`.  Administrator needs to re-run the
-`kubectl run migrate --image=migrator-initializer --restart=OnFailure` command
-to recover.
-
-TODO: it is safe to rollback an apiserver to the previous configuration without
-waiting for the migration to complete. It is only unsafe to roll-forward or
-rollback twice. We need to design how to record the previous configuration.
+complete. 
 
-## Graduation Criteria
+With the newly introduced [storageStates API][], the cluster administrator can
+fast upgrade/downgrade as long as the new apiserver binaries understand all
+versions recorded in storageState.status.persistedStorageVersionHashes.
 
-* alpha: delivering a tool that implements the "alpha workflow" and "failure
-  recovery" sections. ETA is 2018 Q4. 
+[storageStates API]:storage-migration-auto-trigger.md#api-design
 
-* beta: implementing the "beta workflow" and integrating the storage migration
-  into Kubernetes upgrade tests.
+## Beta Graduation Criteria
 
-* GA: TBD.
+* Visibility
+  * metrics for the speed of migration 
+  * metrics for the number of migrated objects
+  * metrics for pending, succeeded, and failed migrations
+  * metrics for the frequency of migrations
 
-We will revisit this section in 2018 Q4. 
+* End-to-end testing
+  * testing migration for CRDs
+  * chaos testing the migrator with injected crashes and network errors
```

Comment:
crashes of both the migrator and API server I would think

Reply:
Added.